starlette/middleware/base.py

```diff
@@ -47,13 +47,23 @@ async def coro() -> None:
         assert message["type"] == "http.response.start"
 
         async def body_stream() -> typing.AsyncGenerator[bytes, None]:
-            while True:
-                message = await queue.get()
-                if message is None:
-                    break
+            # In non-streaming responses, there should be one message to emit
+            message = await queue.get()
+            if message is not None:
                 assert message["type"] == "http.response.body"
                 yield message.get("body", b"")
-            task.result()
+
+                while message.get("more_body"):
```

Comment:
The way I'm seeing it, these changes are related to background task support — i.e. making sure that `body_stream()` can terminate even if `app` is not finished yet _(because it's calling into background tasks, right?)_.

So first question: can we isolate the `maxsize=1` change (+ tests) into its own "fix memory usage" PR? (The title of this PR is already pretty explicit that it's doing "two things" :-))

Second question I'd have, is that for this particular part of the code it seems we could reduce the diff by keeping the previous `while True` loop style - WDYT? Probably something like this…

```diff
        async def body_stream() -> typing.AsyncGenerator[bytes, None]:
            while True:
                message = await queue.get()
                if message is None:
                    break
                assert message["type"] == "http.response.body"
                yield message.get("body", b"")
+               if not message.get("more_body"):
+                   break
+           if task.done():
+               # <Comment>
                task.result()
```

Reply:
When I started reading this code, I found it confusing, so I think my overall goal in addition to fixing the issues mentioned at top was to make the intentions plain and to leave it as readable as possible. I was hoping to help anyone in the future who comes to read this section. (If that's not achieved, then I would like to improve it.) 

Anyway, that was the reason for switching out the `while True` for a loop that's more explicit: I was hoping it would become a lot more obvious that, as Tom said in the chat the other day, it's "almost always the case" that the first message ends the body and that looping is only necessary in the case of `more_body`. This is subjective, of course, so I'm happy to do whatever the group thinks is best. On the whole, though, I would argue that this section is still subtle no matter what direction we go in and that we need to provide some strong hints to future readers what's going on here either in the way the code's organized or via comments or both.

I don't think I have an opinion on the first point you raised. I'm fine with a single PR because the changes are to this single function, `call_next`, but it doesn't matter to me. I can collapse the commits into two to make it simpler to read the history, though, if that's preferred: one for `maxsize=1` and one for background (do these PRs get squash merged)?

Reply:
@erewok

- Sure, thanks for pushing back on my yakshaving 😅 I understand your point about making it easier for new readers to understand the code. To me both ways are clear enough but obviously I've heads down in this for way too long to tell. 😂 So if this new implementation is clearer to you then let's roll with it!
- Yes, I actually think we have squash merge enforced on all PRs, which is mostly why I was encouraging to open two separate PRs. I'm not putting a hard stop on it nor mean to require anyone's time so okay to leave things as is too.